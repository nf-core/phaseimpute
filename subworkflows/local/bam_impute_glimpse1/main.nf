
include { BAM_GL_BCFTOOLS                    } from '../bam_gl_bcftools'
include { GLIMPSE_PHASE                      } from '../../../modules/nf-core/glimpse/phase'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_1 } from '../../../modules/nf-core/bcftools/index'
include { GLIMPSE_LIGATE                     } from '../../../modules/nf-core/glimpse/ligate'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2 } from '../../../modules/nf-core/bcftools/index'

workflow BAM_IMPUTE_GLIMPSE1 {

    take:
    ch_input        // channel (mandatory): [ [id], bam, bai ]
    ch_posfile      // channel (mandatory): [ [panel, chr], legend ]
    ch_panel        // channel (mandatory): [ [panel, chr], vcf, tbi ]
    ch_chunks       // channel  (optional): [ [panel, chr], region1, region2 ]
    ch_fasta        // channel (mandatory): [ [genome], fa, fai ]

    main:

    ch_versions = Channel.empty()
    ch_multiqc_files = Channel.empty()

    // Glimpse1 subworkflow
    BAM_GL_BCFTOOLS( // Compute GL for input data once per panel by chromosome
        ch_input,
        ch_posfile,
        ch_fasta
    )
    ch_multiqc_files = ch_multiqc_files.mix(BAM_GL_BCFTOOLS.out.multiqc_files)
    ch_versions = ch_versions.mix(BAM_GL_BCFTOOLS.out.versions)

    samples_file = Channel.of([[]]).collect()
    gmap_file    = Channel.of([[]]).collect()

    // Combine chunks with panel
    ch_chunks_panel = ch_chunks
        .combine(ch_panel, by:0)
        .map{ metaPC, regionin, regionout, panel, index ->
            [["panel": metaPC.id, "chr": metaPC.chr], regionin, regionout, panel, index]
        }

    // Join input and chunks reference
    ch_phase_input = BAM_GL_BCFTOOLS.out.vcf
        .map{ metaIPC, vcf, index -> [metaIPC.subMap("panel", "chr"), metaIPC, vcf, index] }
        .combine(samples_file)
        .combine(ch_chunks_panel, by: 0)
        .combine(gmap_file)
        .map{ metaPC, metaIPC, bam, bai, samples, regionin, regionout, panel, panel_index, gmap ->
            [metaIPC + ["chunk": regionout],
            bam, bai, samples, regionin, regionout, panel, panel_index, gmap]
        }

    GLIMPSE_PHASE ( ch_phase_input ) // [meta, vcf, index, sample, regionin, regionout, ref, ref_index, map]
    ch_versions = ch_versions.mix(GLIMPSE_PHASE.out.versions )

    BCFTOOLS_INDEX_1 ( GLIMPSE_PHASE.out.phased_variants )
    ch_versions = ch_versions.mix( BCFTOOLS_INDEX_1.out.versions )

    // Ligate all phased files in one and index it
    ligate_input = GLIMPSE_PHASE.out.phased_variants
        .join( BCFTOOLS_INDEX_1.out.csi )
        .map{ metaIPCR, vcf, index -> [metaIPCR.subMap("id", "panel", "chr"), vcf, index] }
        .groupTuple()

    GLIMPSE_LIGATE ( ligate_input )
    ch_versions = ch_versions.mix(GLIMPSE_LIGATE.out.versions )

    BCFTOOLS_INDEX_2 ( GLIMPSE_LIGATE.out.merged_variants )
    ch_versions = ch_versions.mix( BCFTOOLS_INDEX_2.out.versions )

    // Join imputed and index files
    ch_imputed_vcf_tbi = GLIMPSE_LIGATE.out.merged_variants
        .join(BCFTOOLS_INDEX_2.out.tbi)
        .map{ metaIPC, vcf, index -> [metaIPC + [tools: "glimpse1"], vcf, index] }

    emit:
    vcf_tbi             = ch_imputed_vcf_tbi    // channel: [ [id, panel, chr, tool], vcf, tbi ]
    versions            = ch_versions           // channel: [ versions.yml ]
    multiqc_files       = ch_multiqc_files      // channel: [ multiqc_files.yml ]
}
