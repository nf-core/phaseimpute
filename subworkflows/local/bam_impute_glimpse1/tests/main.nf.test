nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_GLIMPSE1"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_IMPUTE_GLIMPSE1"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_glimpse1"
    tag "bam_impute_glimpse1"
    tag "bam_gl_bcftools"

    tag "glimpse"
    tag "glimpse/phase"
    tag "glimpse/ligate"
    tag "bcftools"
    tag "bcftools/index"

    test("Impute with glimpse1") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]).collect()
                input[3] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22:16570065-16597215", "chr22:16570065-16592216"],
                    [[chr: "22", id: "1000GP"], "chr22:16587172-16609999", "chr22:16592229-16609999"]
                )
                input[4] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("Summary") }
            )
        }
    }
}
