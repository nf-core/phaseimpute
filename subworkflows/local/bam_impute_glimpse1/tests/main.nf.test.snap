{
    "Impute with glimpse1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,0cfebd7b9e54867eecaa83a05b28fb73",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,2f4d48a88b06e39ca27e92a80aed8637"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA19401_22_ligate.vcf.gz:md5,bdc87c2bdf3d4ec70716388319d56f5f",
                        "NA19401_22_ligate.vcf.gz.tbi:md5,0c10468e0a63a5d175f23c6bf731ed66"
                    ]
                ],
                "1": [
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,470f38ebb4e2333ff70702df9716dfa2",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,ba880476e98fa9825649bb9d394a96b5",
                    "versions.yml:md5,ba880476e98fa9825649bb9d394a96b5",
                    "versions.yml:md5,cffe7f76c26424afef6177533d1b8768",
                    "versions.yml:md5,cffe7f76c26424afef6177533d1b8768",
                    "versions.yml:md5,e554dad12029fb193020cbcdbc1519cc",
                    "versions.yml:md5,f630355fc3b592f13a2ca2f1edd4aad1",
                    "versions.yml:md5,fe05ac885c0841c53f9f4a1cbae5d9e9"
                ],
                "2": [
                    "NA12878.bcftools_stats.txt:md5,4cacbe4d6dfba4a8035e37611c91923c",
                    "NA19401.bcftools_stats.txt:md5,f91fd586acfdf3845d111a062fbb263f"
                ],
                "multiqc_files": [
                    "NA12878.bcftools_stats.txt:md5,4cacbe4d6dfba4a8035e37611c91923c",
                    "NA19401.bcftools_stats.txt:md5,f91fd586acfdf3845d111a062fbb263f"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,0cfebd7b9e54867eecaa83a05b28fb73",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,2f4d48a88b06e39ca27e92a80aed8637"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA19401_22_ligate.vcf.gz:md5,bdc87c2bdf3d4ec70716388319d56f5f",
                        "NA19401_22_ligate.vcf.gz.tbi:md5,0c10468e0a63a5d175f23c6bf731ed66"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,37a7932cdb5436e6de517224997782b7",
                    "versions.yml:md5,470f38ebb4e2333ff70702df9716dfa2",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,a3db5c2981d19a6f37340bb9c7cede83",
                    "versions.yml:md5,ba880476e98fa9825649bb9d394a96b5",
                    "versions.yml:md5,ba880476e98fa9825649bb9d394a96b5",
                    "versions.yml:md5,cffe7f76c26424afef6177533d1b8768",
                    "versions.yml:md5,cffe7f76c26424afef6177533d1b8768",
                    "versions.yml:md5,e554dad12029fb193020cbcdbc1519cc",
                    "versions.yml:md5,f630355fc3b592f13a2ca2f1edd4aad1",
                    "versions.yml:md5,fe05ac885c0841c53f9f4a1cbae5d9e9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:53:21.253850871"
    },
    "Summary": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:53:21.988387905"
    }
}